docs/gen-ai/CausalLMPipeline.md

```diff
@@ -0,0 +1,118 @@
+# What is a causal language model pipeline?
+
+The causal language model pipeline is a utility class which wraps a tokenizer and a causal language model and provides a uniformed interface for various decoding method to generate text. The pipeline is designed to be easy to use and requires only a few lines of code to generate text.
+
+In Microsoft.ML.GenAI, we will provide a generic `CausalLMPipeline` class plus a typed `CausalLMPipeline` class which specifies the type parameters for the tokenizer and the causal language model. The typed `CausalLMPipeline` class make it easier to develop consuming method for semantic kernel. see [here](./Usage.md#consume-model-from-semantic-kernel) for more details
+# Contract
+```C#
+public abstract class CausalLMPipeline
+{
+    public virtual (
+        Tensor, // output token ids [batch_size, sequence_length]
+        Tensor // output logits [batch_size, sequence_length, vocab_size]
+    ) Generate(
+        Tensor inputIds, // input token ids [batch_size, sequence_length]
+        Tensor attentionMask, // attention mask [batch_size, sequence_length]
+        float temperature = 0.7f,
+        float topP = 0.9f,
+        int maxLen = 128,
+        int[][]? stopTokenSequence = null,
+        bool echo = false); // echo the input token ids in the output token ids
+}
+
+public CasualLMPipeline<TTokenizer, TCausalLM> : CausalLMPipeline
+    where TTokenizer : ITokenizer
+    where TCausalLM : nn.Module<CausalLanguageModelInput, CausalLanguageModelOutput>
+{
+    public CausalLMPipeline<LLama2Tokenizer, Phi3ForCasualLM> Create(LLama2Tokenizer tokenizer, Phi3ForCasualLM model);
```

Comment:
Presumably this is intended to instead be a static method on the base class?

Reply:
Hmmm on the second thought, maybe not. Because `CasualLMPipeline` and `CasualLMPipeline<TTokenizer, TCasualLM>` will live in `GenAI.Core` while `Phi3ForCasualLM` lives in `GenAI.Phi3`.

In that case, this `Create` method is no longer necessary exist, using a constructor to take `TTokenzier` and `TCausalLM` would be easier to understand